golangci_lint fixer mangles during error

<!--
  This is the template for reporting ALE bugs. Make sure you try updating ALE
  to a more recent version before reporting a bug. Look through existing bug
  reports for similar issues before reporting a new one. Don't leave comments
  about new bugs in the comment section for old issues.

  Make sure to try disabling other plugins and trying to repeat your bug before
  reporting it in ALE. Some times problems can arise when two plugins are used
  together, but often your issues might be problems with other plugins.
-->

## Information

**VIM version**

NVIM v0.11.0                                                                                                                                                                                  
Build type: Release  

Operating System: Debian bookworm

## What went wrong

When running `:ALEFix golangci_lint` on a Go file that does not compile, the entire file is replaced with an error message.

## Reproducing the bug

1. Run `:ALEFix golangci_lint` on the following Go file:

```go
package main

func main() {
	asdf
}
```

2. See that the contents of the buffer are replaced with an error message:

```go
main.go:1: : # tmp
./main.go:4:2: undefined: asdf (typecheck)
package main
```

### :ALEInfo
<details>
  <summary>Expand</summary>

 Current Filetype: go
Available Linters: ['bingo', 'cspell', 'gobuild', 'gofmt', 'golangci-lint', 'gopls', 'gosimple', 'gotype', 'govet', 'golangserver', 'revive', 'staticcheck']
   Linter Aliases:
'gobuild' -> ['go build']
'govet' -> ['go vet']
  Enabled Linters: []
  Ignored Linters: []
 Suggested Fixers:
  'gofmt' - Fix Go files with go fmt.
  'gofumpt' - Fix Go files with gofumpt, a stricter go fmt.
  'goimports' - Fix Go files imports with goimports.
  'golangci_lint' - Fix Go files with golangci-lint.
  'golines' - Fix Go file long lines with golines
  'gopls' - Fix Go files with gopls.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
 
 Linter Variables:
" Press Space to read :help for a setting
let g:ale_go_go_executable = 'go'
let g:ale_go_golangci_lint_executable = 'golangci-lint'
let g:ale_go_golangci_lint_options = ''
let g:ale_go_golangci_lint_package = 1
let g:ale_go_langserver_executable = 'go-langserver'
let g:ale_go_langserver_options = ''
 
 Global Variables:
" Press Space to read :help for a setting
let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = v:null
let g:ale_command_wrapper = ''
let g:ale_completion_delay = v:null
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = v:null
let g:ale_disable_lsp = v:true
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = v:null
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = []
let g:ale_history_enabled = 1
let g:ale_info_default_mode = 'preview'
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = v:null
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = v:false
let g:ale_lint_on_save = v:true
let g:ale_lint_on_text_changed = 'never'
let g:ale_linter_aliases = {}
let g:ale_linters = {'sh': ['shell', 'shellcheck']}
let g:ale_linters_explicit = v:true
let g:ale_linters_ignore = {}
let g:ale_list_vertical = v:null
let g:ale_list_window_size = v:null
let g:ale_loclist_msg_format = v:null
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = v:null
let g:ale_maximum_file_size = v:null
let g:ale_open_list = v:null
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_root = {}
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = v:false
let g:ale_set_quickfix = v:true
let g:ale_set_signs = 1
let g:ale_sign_column_always = v:null
let g:ale_sign_error = v:null
let g:ale_sign_info = v:null
let g:ale_sign_offset = v:null
let g:ale_sign_style_error = v:null
let g:ale_sign_style_warning = v:null
let g:ale_sign_warning = v:null
let g:ale_sign_highlight_linenrs = v:null
let g:ale_type_map = v:null
let g:ale_use_neovim_diagnostics_api = 1
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 'all'
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
 
  Command History:

(finished - exit code 1) ['/bin/bash', '-c', '''golangci-lint'' run  --fix < ''/tmp/nvim.cbandy/lnkZmr/0/main.go''']

</details>

--- a/autoload/ale/fixers/golangci_lint.vim
+++ b/autoload/ale/fixers/golangci_lint.vim
@@ -1,32 +1,48 @@
 " Author: Ian Stapleton Cordasco <graffatcolmingov@gmail.com>
 " Description: Run golangci-lint with the --fix flag to autofix some issues
 
-call ale#Set('go_golangci_lint_options', '')
-call ale#Set('go_golangci_lint_executable', 'golangci-lint')
-call ale#Set('go_golangci_lint_package', 1)
+call ale#Set('go_golangci_formatter_options', '')
+call ale#Set('go_golangci_formatter_executable', 'golangci-lint')
 
-function! ale#fixers#golangci_lint#GetCommand(buffer) abort
+function! ale#fixers#golangci_lint#GetExecutable(buffer) abort
+    let l:executable = ale#Var(a:buffer, 'go_golangci_formatter_executable')
+
+    return l:executable
+endfunction
+
+function! ale#fixers#golangci_lint#GetCommand(buffer, version) abort
     let l:filename = expand('#' . a:buffer . ':t')
-    let l:executable = ale#Var(a:buffer, 'go_golangci_lint_executable')
-    let l:options = ale#Var(a:buffer, 'go_golangci_lint_options') . ' --fix'
-    let l:package_mode = ale#Var(a:buffer, 'go_golangci_lint_package')
+    let l:executable = ale#fixers#golangci_lint#GetExecutable(a:buffer)
+    let l:options = ale#Var(a:buffer, 'go_golangci_formatter_options')
     let l:env = ale#go#EnvString(a:buffer)
 
-
-    if l:package_mode
+    if ale#semver#GTE(a:version, [2, 0, 0])
         return l:env . ale#Escape(l:executable)
-        \   . ' run '
-        \   .  l:options
+        \   . ' fmt --stdin '
+        \   . l:options
+    else
+        return l:env . ale#Escape(l:executable)
+        \   . ' run --fix '
+        \   . l:options
+        \   . ' '
+        \   . ale#Escape(l:filename)
     endif
+endfunction
 
-    return l:env . ale#Escape(l:executable)
-    \   . ' run '
-    \   . l:options
-    \   . ' ' . ale#Escape(l:filename)
+function! ale#fixers#golangci_lint#GetCommandForVersion(buffer, version) abort
+    return {
+    \ 'command': ale#fixers#golangci_lint#GetCommand(a:buffer, a:version)
+    \}
 endfunction
 
 function! ale#fixers#golangci_lint#Fix(buffer) abort
-    return {
-    \  'command': ale#fixers#golangci_lint#GetCommand(a:buffer),
-    \}
+    let l:executable =  ale#fixers#golangci_lint#GetExecutable(a:buffer)
+    let l:command = ale#fixers#golangci_lint#GetExecutable(a:buffer) . ale#Pad('--version')
+
+    return ale#semver#RunWithVersionCheck(
+    \   a:buffer,
+    \   l:executable,
+    \   l:command,
+    \   function('ale#fixers#golangci_lint#GetCommandForVersion'),
+    \)
 endfunction

--- a/doc/ale-go.txt
+++ b/doc/ale-go.txt
@@ -144,7 +144,7 @@ g:ale_go_golangci_lint_options
   Default: `''`
 
   This variable can be changed to alter the command-line arguments to the
-  golangci-lint invocation.
+  golangci-lint run invocation.
 
                                          *ale-options.go_golangci_lint_package*
                                                *g:ale_go_golangci_lint_package*
@@ -157,6 +157,30 @@ g:ale_go_golangci_lint_package
   When set to `1`, the whole Go package will be checked instead of only the
   current file.
 
+golangci_lint can also be user as a fixer to format go source files. In this
+case the following configuration variables can be used to configure the
+formatters:
+
+                                    *ale-options.go_golangci_formatter_executable*
+                                          *g:ale_go_golangci_formatter_executable*
+                                          *b:ale_go_golangci_formatter_executable*
+go_golangci_formatter_executable
+g:ale_go_golangci_formatter_executable
+  Type: |String|
+  Default: `'golangci-lint'`
+
+  The executable that will be run for golangci-lint.
+
+                                      *ale-options.go_golangci_formatter_options*
+                                            *g:ale_go_golangci_formatter_options*
+                                            *b:ale_go_golangci_formatter_options*
+go_golangci_formatter_options
+g:ale_go_golangci_formatter_options
+  Type: |String|
+  Default: `''`
+
+  This variable can be changed to alter the command-line arguments to the
+  golangci-lint fmt invocation.
 
 ===============================================================================
 golangserver                                              *ale-go-golangserver*

--- a/test/fixers/test_golangci_lint_fixer_callback.vader
+++ b/test/fixers/test_golangci_lint_fixer_callback.vader
@@ -1,48 +1,80 @@
 Before:
+  call ale#assert#SetUpFixerTest('go', 'golangci_lint')
   Save g:ale_go_go111module
-  Save g:ale_go_golangci_lint_executable
-  Save g:ale_go_golangci_lint_options
-  Save g:ale_go_golangci_lint_package
+  Save g:ale_go_golangci_formatter_executable
+  Save g:ale_go_golangci_formatter_options
 
-  " Use an invalid global executable, so we don't match it.
-  let g:ale_go_golangci_lint_executable = 'xxxinvalid'
-  let g:ale_go_golangci_lint_options = ''
 
   call ale#test#SetDirectory('/testplugin/test/fixers')
   call ale#test#SetFilename('../test-files/go/testfile.go')
+
 After:
-  Restore
+  call ale#test#RestoreDirectory()
+  call ale#assert#TearDownFixerTest()
 
   unlet! b:ale_go_go111module
 
-  call ale#test#RestoreDirectory()
+Execute(The golangci-lint callback should return the correct default values with v1):
 
-Execute(The golangci-lint callback should return the correct default values):
+  GivenCommandOutput ['golangci-lint has version 1.64.8 built with go1.23.0']
 
-  AssertEqual
+  AssertFixer
   \ {
-  \   'command': ale#Escape('xxxinvalid') . ' run  --fix',
+  \   'command': ale#Escape('golangci-lint') . ' run --fix  ' . ale#Escape('testfile.go'),
+  \ }
+
+Execute(The golangci-lint callback should include custom golangci-lint options with v1):
+  let g:ale_go_golangci_formatter_options = "--new --config /dev/null"
+
+  GivenCommandOutput ['golangci-lint has version 1.64.8 built with go1.23.0']
+
+  AssertFixer
+  \ {
+  \   'command': ale#Escape('golangci-lint')
+  \     . ' run --fix ' . g:ale_go_golangci_formatter_options . ' ' . ale#Escape('testfile.go'),
   \ },
-  \ ale#fixers#golangci_lint#Fix(bufnr(''))
 
-Execute(The golangci-lint callback should include custom golangci-lint options):
-  let g:ale_go_golangci_lint_options = "--new --config /dev/null"
+Execute(The golangci-lint callback should override executable with v1):
+  let g:ale_go_golangci_formatter_executable = 'xxxinvalid'
 
-  AssertEqual
+  GivenCommandOutput ['golangci-lint has version 1.64.8 built with go1.23.0']
+
+  AssertFixer
   \ {
   \   'command': ale#Escape('xxxinvalid')
-  \     . ' run ' . g:ale_go_golangci_lint_options . ' --fix',
+  \     . ' run --fix '
+  \     . g:ale_go_golangci_formatter_options
+  \     . ' ' . ale#Escape('testfile.go'),
   \ },
-  \ ale#fixers#golangci_lint#Fix(bufnr(''))
 
-Execute(The golangci-lint callback should support per-file mode):
-  let g:ale_go_golangci_lint_package = 0
+Execute(The golangci-lint callback should return the correct default values with v2):
 
-  AssertEqual
+  GivenCommandOutput ['golangci-lint has version 2.1.5 built with go1.23.0']
+
+  AssertFixer
+  \ {
+  \   'command': ale#Escape('golangci-lint') . ' fmt --stdin ',
+  \ }
+
+Execute(The golangci-lint callback should include custom golangci-lint options with v2):
+  let g:ale_go_golangci_formatter_options = "--new --config /dev/null"
+
+  GivenCommandOutput ['golangci-lint has version 2.1.5 built with go1.23.0']
+
+  AssertFixer
+  \ {
+  \   'command': ale#Escape('golangci-lint')
+  \     . ' fmt --stdin ' . g:ale_go_golangci_formatter_options,
+  \ },
+
+Execute(The golangci-lint callback should override executable with v2):
+  let g:ale_go_golangci_formatter_executable = 'xxxinvalid'
+
+  GivenCommandOutput ['golangci-lint has version 2.1.5 built with go1.23.0']
+
+  AssertFixer
   \ {
   \   'command': ale#Escape('xxxinvalid')
-  \     . ' run '
-  \     . g:ale_go_golangci_lint_options
-  \     . ' --fix ' . ale#Escape('testfile.go'),
+  \     . ' fmt --stdin '
+  \     . g:ale_go_golangci_formatter_options
   \ },
-  \ ale#fixers#golangci_lint#Fix(bufnr(''))

